### DISCUSSION TITLE
Suggestions for interactive API documentation

### DISCUSSION THREAD

**Discussion by jmdye** [mkdocs/mkdocs#3864] [Q&A]
Looking for a way to create interactive API documentation. Many of us have probably seen this before in solutions such as [redocly](https://redocly.com), [scalar](https://scalar.com) and so many others but all of these solutions have costs. Does anyone know if there any plugins or other solutions anyone knows of that would be relatively simple to integrate with a MkDocs site where one could simply reference an OpenAPI specification in the nav.yml and the plugin or whatever would recognize that and create the interactive elements so that doc users could more interactively explore an API's capabilities within the docs? I didn't see anything of this nature in the MkDocs extensions or plugins pages/repos.

**Comment by pawamoy** [comment]
In the [catalog](https://github.com/mkdocs/catalog), there are these two plugins:

- https://github.com/Neoteroi/mkdocs-plugins
- https://github.com/bharel/mkdocs-render-swagger-plugin

**Reply by Andre601** [reply]
I personally are in favor of the first as it integrates more into the style of the docs where the second is pretty just an iframe being inserted (Edit: I thought of a different plugin).

Tho, I personally have a hook for more customization used:
https://github.com/purrbot-site/Docs/blob/master/theme/.hooks/api_docs_gen.py

**Comment by jmdye** [comment]
Thank you @pawamoy & @Andre601. More specifically, I was looking for an easy way to implement the more interactive components, rather than just rendering the static content. I really like how some of these solutions generate API usage examples in multiple languages, and even allow those examples to be executed - just based on the swagger doc. A plugin or framework I might be able to easily implement to do something like this is what I am really after.
![Screenshot 2024-11-18 at 10 51 55 AM](https://github.com/user-attachments/assets/90007480-2b1f-409d-a727-634b3542fb4c)

**Reply by Andre601** [reply]
There is the swagger-ui-tag plugin, which inserts an iframe with the Swagger UI as you have by default, which also would include options to test the endpoints directly.

Biggest downside is tge fact it's just a iframe, so it may not be as customizable in terms of appearance as other solutions.

**Reply by kamilkrzyskow** [reply]
MkDocs mostly focuses on Python, the back-end generation logic of the static HTML pages. Even the `material` theme doesn't really implement JavaScript to inject content or components, content generation is Python based. There was a similar discussion about MDX and the usage of JSX components, which had a similar answer:
- https://github.com/squidfunk/mkdocs-material/discussions/3621

The thing is that the DOM structure, the classes etc. are defined by the theme. So even if somebody would implement this, then it would be compatible with only one specific theme and HTML structure, which isn't ideal in the MkDocs ecosystem that promotes a general compatibility. 

So like Andre is saying this feature (the interactivity) would likely need to be handled via some external service, and put onto the page via some iframe tag.

**Reply by pawamoy** [reply]
That's actually what I did in a previous job: injecting the FastAPI-generated Swagger docs in the page thanks to an iframe. It worked well (given CORS are correctly setup).

Execution of code blocks will be even more difficult (well, Javascript will work, there's Pyodide for Python, but other languages might be difficult to execute client-side).
